opal/mca/btl/tcp/btl_tcp_endpoint.c

```diff
@@ -717,7 +717,35 @@ static int mca_btl_tcp_endpoint_start_connect(mca_btl_base_endpoint_t* btl_endpo
 
     /* start the connect - will likely fail with EINPROGRESS */
     mca_btl_tcp_proc_tosocks(btl_endpoint->endpoint_addr, &endpoint_addr);
-
+    
+    /* Bind the socket to one of the addresses associated with
+     * this btl module.  This sets the source IP to one of the 
+     * addresses shared in modex, so that the destination rank 
+     * can properly pair btl modules, even in cases where Linux 
+     * might do something unexpected with routing */
+    opal_socklen_t sockaddr_addrlen = sizeof(struct sockaddr_storage);
+    if (endpoint_addr.ss_family == AF_INET) {
```

Comment:
Minor nit, but I would change this to have a struct sockaddr* temporary variable with assignment from either tcp_ifaddr or tcp_ifaddr6, to avoid duplicating the bind() and error handling code.